File: src/python/xerxes/auth/oauth.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Provider-agnostic OAuth helpers.

Wraps the MCP OAuth helpers with a thin ``OAuthClient`` that targets
common providers (OpenAI, Anthropic, GitHub Copilot, GitHub PAT).
Each provider sets a different ``OAuthConfig``; the client object
exposes ``begin_authorize``, ``finish_authorize``, and ``refresh``.
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass

import httpx

from ..mcp.oauth import OAuthConfig as McpOAuthConfig
from ..mcp.oauth import OAuthToken, generate_pkce_pair
from ..mcp.oauth import build_authorize_url as _build_url
from ..mcp.oauth import exchange_code as _exchange
from ..mcp.oauth import refresh_token as _refresh


@dataclass
class OAuthConfig(McpOAuthConfig):
    """Provider configuration alias of :class:`mcp.oauth.OAuthConfig`."""

    pass


@dataclass
class _AuthorizeContext:
    """Material returned by :meth:`OAuthClient.begin_authorize`.

    Attributes:
        url: Authorize URL to open in the user's browser.
        state: Anti-CSRF token; must be echoed in the callback.
        code_verifier: PKCE verifier required to exchange the code.
    """

    url: str
    state: str
    code_verifier: str


class OAuthClient:
    """Authorization-code-with-PKCE client bound to one provider config."""

    def __init__(self, config: OAuthConfig, *, http_client: httpx.Client | None = None) -> None:
        """Bind to ``config``; reuse ``http_client`` if supplied."""
        self._config = config
        self._http = http_client

    def begin_authorize(self) -> _AuthorizeContext:
        """Start the flow: generate PKCE pair and the authorize URL."""
        verifier, challenge = generate_pkce_pair()
        state = secrets.token_urlsafe(24)
        url = _build_url(self._config, state=state, code_challenge=challenge)
        return _AuthorizeContext(url=url, state=state, code_verifier=verifier)

    def finish_authorize(self, *, code: str, code_verifier: str) -> OAuthToken:
        """Exchange the redirect-supplied ``code`` for an access token."""
        return _exchange(self._config, code=code, code_verifier=code_verifier, client=self._http)

    def refresh(self, token: OAuthToken) -> OAuthToken:
        """Refresh ``token`` using its refresh grant, returning a new token."""
        return _refresh(self._config, token=token, client=self._http)


# ---------------------------- provider presets -----------------------------


def github_pat_preset(client_id: str) -> OAuthConfig:
    """OAuth config for the standard GitHub user-token flow."""
    return OAuthConfig(
        client_id=client_id,
        authorize_url="https://github.com/login/oauth/authorize",
        token_url="https://github.com/login/oauth/access_token",
        scopes=("read:user",),
    )


def openai_preset(client_id: str) -> OAuthConfig:
    """OAuth config for OpenAI's auth-code flow."""
    return OAuthConfig(
        client_id=client_id,
        authorize_url="https://auth.openai.com/oauth/authorize",
        token_url="https://auth.openai.com/oauth/token",
        scopes=("openid", "profile"),
    )


def anthropic_preset(client_id: str) -> OAuthConfig:
    """OAuth config for the Anthropic / Claude Code creds bridge."""
    return OAuthConfig(
        client_id=client_id,
        authorize_url="https://console.anthropic.com/oauth/authorize",
        token_url="https://console.anthropic.com/oauth/token",
        scopes=("openid",),
    )


def copilot_preset(client_id: str) -> OAuthConfig:
    """OAuth config for GitHub Copilot bearer-token issuance."""
    return OAuthConfig(
        client_id=client_id,
        authorize_url="https://github.com/login/oauth/authorize",
        token_url="https://github.com/login/oauth/access_token",
        scopes=("copilot",),
    )


__all__ = [
    "OAuthClient",
    "OAuthConfig",
    "OAuthToken",
    "anthropic_preset",
    "copilot_preset",
    "github_pat_preset",
    "openai_preset",
]

File: src/typescript/src/auth/oauth.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { randomBytes } from 'node:crypto'

import {
  buildAuthorizeUrl,
  exchangeCode,
  generatePkcePair,
  refreshToken,
  type OAuthConfig,
  type OAuthFetch,
  type OAuthFetchOptions,
  OAuthToken,
} from '../mcp/oauth.js'

export { OAuthToken }
export type { OAuthConfig }

export interface AuthorizeContext {
  readonly codeVerifier: string
  readonly state: string
  readonly url: string
}

export interface OAuthClientOptions {
  readonly fetchImplementation?: OAuthFetch
  readonly now?: () => number
  readonly stateGenerator?: () => string
}

/** Authorization-code-with-PKCE client bound to one provider configuration. */
export class OAuthClient {
  readonly config: OAuthConfig

  private readonly fetchImplementation: OAuthFetch | undefined
  private readonly now: (() => number) | undefined
  private readonly stateGenerator: () => string

  constructor(config: OAuthConfig, options: OAuthClientOptions = {}) {
    this.config = config
    this.fetchImplementation = options.fetchImplementation
    this.now = options.now
    this.stateGenerator = options.stateGenerator ?? defaultState
  }

  /** Start an OAuth flow and return the browser URL plus callback-verification material. */
  beginAuthorize(): AuthorizeContext {
    const { verifier, challenge } = generatePkcePair()
    const state = this.stateGenerator()
    if (!state) {
      throw new Error('OAuth state generator returned an empty state')
    }
    return {
      url: buildAuthorizeUrl(this.config, { state, codeChallenge: challenge }),
      state,
      codeVerifier: verifier,
    }
  }

  /** Exchange an authorization code after the caller has verified its callback state. */
  async finishAuthorize(code: string, codeVerifier: string, signal?: AbortSignal): Promise<OAuthToken> {
    return exchangeCode(this.config, {
      code,
      codeVerifier,
      ...this.requestOptions(signal),
    })
  }

  /** Refresh a provider token. */
  async refresh(token: OAuthToken, signal?: AbortSignal): Promise<OAuthToken> {
    return refreshToken(this.config, token, this.requestOptions(signal))
  }

  private requestOptions(signal?: AbortSignal): OAuthFetchOptions {
    return {
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
      ...(this.now ? { now: this.now } : {}),
      ...(signal ? { signal } : {}),
    }
  }
}

/** OAuth configuration for the standard GitHub user-token flow. */
export function githubPatPreset(clientId: string): OAuthConfig {
  return {
    clientId,
    authorizeUrl: 'https://github.com/login/oauth/authorize',
    tokenUrl: 'https://github.com/login/oauth/access_token',
    scopes: ['read:user'],
  }
}

/** OAuth configuration for OpenAI's authorization-code flow. */
export function openaiPreset(clientId: string): OAuthConfig {
  return {
    clientId,
    authorizeUrl: 'https://auth.openai.com/oauth/authorize',
    tokenUrl: 'https://auth.openai.com/oauth/token',
    scopes: ['openid', 'profile'],
  }
}

/** OAuth configuration for Anthropic / Claude Code credential bridging. */
export function anthropicPreset(clientId: string): OAuthConfig {
  return {
    clientId,
    authorizeUrl: 'https://console.anthropic.com/oauth/authorize',
    tokenUrl: 'https://console.anthropic.com/oauth/token',
    scopes: ['openid'],
  }
}

/** OAuth configuration for GitHub Copilot bearer-token issuance. */
export function copilotPreset(clientId: string): OAuthConfig {
  return {
    clientId,
    authorizeUrl: 'https://github.com/login/oauth/authorize',
    tokenUrl: 'https://github.com/login/oauth/access_token',
    scopes: ['copilot'],
  }
}

function defaultState(): string {
  return Buffer.from(randomBytes(24)).toString('base64url')
}
